src/compiler/visitorPublic.ts

```diff
@@ -22,35 +22,61 @@ import {
 /**
  * Visits a Node using the supplied visitor, possibly returning a new Node in its place.
  *
+ * - If the input node is undefined, then the output is undefined.
+ * - If the visitor returns undefined, then the output is undefined.
+ * - If the output node is not undefined, then it will satisfy the test function.
+ * - In order to obtain a return type that is more specific than `Node`, a test
+ *   function _must_ be provided, and that function must be a type predicate.
+ *
  * @param node The Node to visit.
  * @param visitor The callback used to visit the Node.
  * @param test A callback to execute to verify the Node is valid.
  * @param lift An optional callback to execute to lift a NodeArray into a valid Node.
  */
-export function visitNode<T extends Node>(node: T, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T;
-
+export function visitNode<TIn extends Node | undefined, TVisited extends Node | undefined, TOut extends Node>(

```

Comment:
When `visitNode` is called with a `test` type guard, `TOut` should be the same as, or a subtype of, `TIn`. If there are places where this doesn't hold true, then those are likely bugs.

We don't check the input node against `test` for performance reasons since we assume that `node` and `test` are aligned.

Reply:
I will have to test this, but my impression was that this didn't work thanks to branded node types.

Reply:
I tested by changing `TOut` to be `TOut extends NonNullable<TIn>`, and it didn't work because `Expression`; the `isExpression` helper ends up not satisfying the signature that requires a test.

I'm not sure that it should be a subtype, though. Take this case:

```ts
    [SyntaxKind.PrefixUnaryExpression]: function visitEachChildOfPrefixUnaryExpression(node, visitor, context, _nodesVisitor, nodeVisitor, _tokenVisitor) {
        return context.factory.updatePrefixUnaryExpression(node,
            Debug.checkDefined(nodeVisitor(node.operand, visitor, isExpression)));
    },
```

`node.operand` is a `UnaryExpression`. It may be the case that a visitor function takes this node in and returns a different kind of expression, and that is accepted because `updatePrefixUnaryExpression` accepts `Expression`. So, I can't fix this by changing `isExpression` to `isUnaryExpression` because that is too narrow.

Reply:
This is a case of `node.operand` having a higher inference priority than the callback. We could avoid that by casting `node.operand` to `Expression | undefined`.

Reply:
I made the change, but it required adding casts to about 70 other calls and didn't find any bugs.

I'm sort of not liking this just because the error isn't very helpful; either it infers the wrong thing and the error happens at use of the result later, or it picks the non-test overload because the types mismatch (in the case of Expression, Statement).

This is definitely something that seems like it'd be annoying for API users, having to cast things that should be assignable but then don't work.

Reply:
I'll also say that my version of these functions _always_ performs the checks. I am currently perf testing to see the impact of these variations, though.

Reply:
The benchmarks we show seem to imply that these don't really matter. This makes sense to me because debug asserts are totally disabled in the benchmarks because `NODE_ENV` is not `development`.

Reply:
Revisiting this, my impression is that it's desirable to not have them be related because the input may not actually be related to the output.

Here's an example from the module transform:

```ts
let expression: Expression = visitNode(node.operand, visitor, isExpression);
if (isPrefixUnaryExpression(node)) {
    expression = factory.updatePrefixUnaryExpression(node, expression);
}
else {
    expression = factory.updatePostfixUnaryExpression(node, expression);
    if (!valueIsDiscarded) {
        temp = factory.createTempVariable(hoistVariableDeclaration);
        expression = factory.createAssignment(temp, expression);
        setTextRange(expression, node);
    }
    expression = factory.createComma(expression, factory.cloneNode(node.operand));
    setTextRange(expression, node);
}
```

Here, `node.operand` is an `Identifier`, however, we're building an expression out of it, and our visitor may transform it to something that isn't an `Identifier`; if we required that `TOut extends TIn`, this code wouldn't be legal.